src/groupchat_packets.md

```diff
@@ -0,0 +1,1128 @@
+
+The procedure of invite is:
+
+- find a group using group_number.
+- if the group doesn't exist
+    - return error.
+- else if the group status is not `connected`
+    - return error.
+- else
+    - assemble and send `Invite packet` using messenger and net crypto.
+    
+When a peer receives `Invite packet` then:
+
+- check if the packet length is correct, if it is not correct
+    - return error.
+- else
+    - get group number using received packet.
+    - join the group
+        - send `Response packet` to the peer.
+        - add friend connection to the group.
+        - send peer query packet.
+
+When a peer receives `Response packet` then:
+
+- check if the packet size is correct, if it is not correct
+    - return error.
+- else
+    - find group using received packet.
+    - if can't find group
+        - return error.
+    - else if group type is not equal to group of us
+        - return error.
+    - else if the group id is not correct
+        - return error.
+    - else
+        - get random peer number which is not redundant.
+        - add the peer to group chat.
+        - add the friend connection to group chat.
+        - send new peer message to all of the group members.
+        
+## Peer online packet
+
+As soon as the connection to the other peer is opened, a `peer online packet` is sent to the peer.
+The purpose of this packet is to tell the peer that we want to establish the group chat connection.
+
+Serialized form:
+
+Length    | Content
+--------- | ------
+`1`       | `0x61`
+`2`       | `group number(local)`
+`1`       | `group type`(0: text, 1: audio)
+`32`      | `unique id`
+
+When a peer receives this packet then:
+
+- get group number of us using received packet's `unique id`.
+- get the group number of the peer us which is in the packet's `group number(local)`.
+- if the peer is already online then
+    - return.
+- else
+    - if the number of close connected peers are zero or the peer is the one we are introducing to conference
+        - send `peer query packet`.
+    - send `peer online packet` using net crypto.
+    - set the status of peer to online.
+    - if the peer is the one who we are introducing
+        - send new peer message to all of the group members.
+    - send ping.
+
+## Peer leave packet
+
+This packet is sent to the peer right before killing a group connection.
+
+Serialized form:
+
+Length    | Content
+--------- | ------
+`1`       | `0x62`
+`2`       | `group number(local)`
+`1`       | `0x01`
+
+When a peer receives this packet then:
+
+- kill the friend connection.
+- set close list to nothing connected.
+
+## Peer query packet
+
+Sent to peer to query the peer list in a group chat. A peer receives this packet send response packet with `response packet`.
+If there are many peers in a group chat, then response packet may larger than the maximum packet size of friend connection packet(1373 bytes),
+then multiple response packets will be sent.
+
+#### Request
```

Comment:
> Request | Response

Does it mean we have 2 different packets under the same id?

Reply:
Yes.
But 4th bytes is different.
```
#### Request

Serialized form:

Length    | Content
--------- | ------
`1`       | `0x62`
`2`       | `group number`
`1`       | `0x08`
```

```
Length    | Content
--------- | ------
`1`       | `0x62`
`2`       | `group number`
`1`       | `0x09`
```